### PR TITLE
Lock exception leaves document with bad revision

### PR DESCRIPTION
**Pull request by mlodick**
A document's local version should only be incremented when the database version is, otherwise, if the lock exception is caught, a second persist & can succeed and overwrite data.

### CODE REVIEW

**Comment by alcaeus** [comment]
Changes look good, can you create a test for this use-case? Thanks!

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 1]
```diff
@@ -426,6 +425,8 @@ public function update($document, array $options = array())
 
             if (($this->class->isVersioned || $this->class->isLockable) && ! $result['n']) {
                 throw LockException::lockFailed($document);
+            } else if ($this->class->isVersioned && $nextVersion) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: We tend to keep `elseif` written together. Also If the class is versioned the `$nextVersion` must be present so there's no need for checking it here (the latter may not be true by looking only at `if` statements you've modifed above but that's validated on [class metadata](https://github.com/doctrine/mongodb-odm/blob/68d4774c09de387a70dc492951f6e38df80ff092/lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php#L1971) level)

Comment (mlodick) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: fair enough.

**Comment by malarzm** [comment]
@mlodick any chance you could write a test for the change you've implemented?

**Comment by mlodick** [comment]
@malarzm sorry, I've been out of town. I'll get a test written today or tomorrow

**Comment by alcaeus** [comment]
@mlodick could you add that test? I'd like to get this merged next week so we can release it in 1.1.6, otherwise I'd have to push it back to 1.1.7.

**Comment by mlodick** [comment]
@alcaeus are these tests ok?

**Comment by alcaeus** [comment]
@mlodick I'll take a look tonight and update them myself if necessary. Thank you very much! ❤️

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php [hunk 2]
```diff
@@ -466,6 +466,55 @@ public function testDefaultWriteConcernIsRespected()
         $this->dm->persist($testDocument);
         $this->dm->flush();
     }
+
+    public function testVersionIncrementOnUpdateSuccess()
+    {
+        $class = DocumentPersisterTestDocumentWithVersion::class;
+        $documentPersister = $this->uow->getDocumentPersister($class);
+
+        $collection = $this->createMock('\MongoCollection');
```
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php]: You can use `MongoCollection::class`

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php [hunk 3]
```diff
@@ -466,6 +466,55 @@ public function testDefaultWriteConcernIsRespected()
...
+            ->will($this->returnValue(array('n' => 0)));
+
+        $reflectionProperty = new \ReflectionProperty($documentPersister, 'collection');
+        $reflectionProperty->setAccessible(true);
+        $reflectionProperty->setValue($documentPersister, $collection);
+
+        $testDocument = new $class();
+        $testDocument->id = 12345;
+        $this->uow->registerManaged($testDocument, 12345, array('id' => 12345));
+        $this->setExpectedException('\Doctrine\ODM\MongoDB\LockException');
```
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php]: `::class`

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php [hunk 4]
```diff
@@ -466,6 +466,55 @@ public function testDefaultWriteConcernIsRespected()
...
+
+        $this->assertEquals(2, $testDocument->revision);
+    }
+
+    public function testNoVersionIncrementOnUpdateFailure()
+    {
+        $class = DocumentPersisterTestDocumentWithVersion::class;
+        $documentPersister = $this->uow->getDocumentPersister($class);
+
+        $collection = $this->createMock('\MongoCollection');
```
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php]: `::class`

**Comment by mlodick** [comment]
@alcaeus I think this is all set then, though I'm not sure how to retrigger the test that failed - seems to have failed getting a pecl package.

**Comment by malarzm** [comment]
@mlodick just restarted the job, let's see how it ends :)

**Comment by malarzm** [comment]
Green it is, thanks @mlodick!

**Comment by mlodick** [comment]
👍  thank you

**Pull request merged**